azure-kusto-ingest/src/resource_manager/ingest_client_resources.rs

```diff
@@ -0,0 +1,148 @@
+use std::sync::Arc;
+
+use crate::client_options::QueuedIngestClientOptions;
+
+use super::{
+    cache::{Cached, Refreshing},
+    resource_uri::{ClientFromResourceUri, ResourceUri},
+    RESOURCE_REFRESH_PERIOD,
+};
+use anyhow::Result;
+use azure_core::ClientOptions;
+use azure_kusto_data::{models::TableV1, prelude::KustoClient};
+use azure_storage_blobs::prelude::ContainerClient;
+use azure_storage_queues::QueueClient;
+use tokio::sync::RwLock;
+
+/// Helper to get a column index from a table
+// TODO: this could be moved upstream into Kusto Data - would likely result in a change to the API of this function to return an Option<usize>
+fn get_column_index(table: &TableV1, column_name: &str) -> Result<usize> {
+    table
+        .columns
+        .iter()
+        .position(|c| c.column_name == column_name)
+        .ok_or(anyhow::anyhow!(
+            "{} column is missing in the table",
+            column_name
+        ))
+}
+
+/// Helper to get a resource URI from a table, erroring if there are no resources of the given name
+fn get_resource_by_name(table: &TableV1, resource_name: String) -> Result<Vec<ResourceUri>> {
+    let storage_root_index = get_column_index(table, "StorageRoot")?;
+    let resource_type_name_index = get_column_index(table, "ResourceTypeName")?;
+
+    let resource_uris: Vec<Result<ResourceUri>> = table
+        .rows
+        .iter()
+        .filter(|r| r[resource_type_name_index] == resource_name)
+        .map(|r| {
+            ResourceUri::try_from(r[storage_root_index].as_str().ok_or(anyhow::anyhow!(
+                "Response returned from Kusto could not be parsed as a string"
+            ))?)
+        })
+        .collect();
+
+    if resource_uris.is_empty() {
+        return Err(anyhow::anyhow!(
+            "No {} resources found in the table",
+            resource_name
+        ));
+    }
+
+    resource_uris.into_iter().collect()
+}
+
+/// Helper to turn a vector of resource URIs into a vector of Azure clients of type T with the provided [ClientOptions]
+fn create_clients_vec<T>(resource_uris: &[ResourceUri], client_options: &ClientOptions) -> Vec<T>
+where
+    T: ClientFromResourceUri,
+{
+    resource_uris
+        .iter()
+        .map(|uri| T::create_client(uri.clone(), client_options.clone()))
+        .collect()
+}
+
+/// Storage of the clients required for ingestion
+#[derive(Debug, Clone)]
+pub struct InnerIngestClientResources {
+    pub secured_ready_for_aggregation_queues: Vec<QueueClient>,
```

Comment:
naming: This seems an overly complex name!  Perhaps simply call it `ingestion_queues` or `ingestion_queue_clients`?